### DISCUSSION TITLE
Try and handling (specific) typed exceptions

### DISCUSSION THREAD

**Discussion by NickDarvey** [louthy/language-ext#1209] [Q&A]
I'm wrapping a library which throws on expected errors.

```csharp
private EitherAsync<AccessFailure, AuthenticationResult> TryGetAccessToken (IAccount account) =>
  TryAsync(_pca.AcquireTokenSilent(_scopes, account).ExecuteAsync())
    .Map(Either<AccessFailure, AuthenticationResult>.Right)
    .IfFail()
    .With<MsalUiRequiredException>(_ => Left(AccessFailure.AccessDenied))
    .OtherwiseReThrow()
    .ToAsync();
```

Is there a way to avoid the explicit types in `.Map(Either<AccessFailure, AuthenticationResult>.Right)`?
(If I just use `Right` method from prelude then it returns an `EitherRight` to which my `Left` is not convertible.)

Otherwise, I'm sure this is a common pattern (going from Trys to Eithers), so is there a better way to approach it altogether? (I had a look in the tests but didn't examples which match on specific exception types.)

**Comment by louthy** [comment]
There’s two natural-transformation extension methods called ToEither:


One maps to Either<Exception, A>, the other [allows you to map the Excepton](https://github.com/louthy/language-ext/blob/main/LanguageExt.Core/Monads/Alternative%20Value%20Monads/Try/Try/Try.Extensions.cs#L294)

**Reply by NickDarvey** [reply]
RE: mapping the exception, is that just a matter of:

```csharp
TryAsync(etc)
  .ToEither(e =>
    e is MyFirstException => MyFirstError ?
    : e is MySecondException => MySecondError ?
    : ExceptionDispatchInfo.Capture(e).Throw())
```
or is there a way of getting an ExceptionMatch helper here too?

Also the TryAsync variant [wraps it in an `Common.Error`](https://github.com/louthy/language-ext/blob/437171250d26a5b26c516b7c0ca94ebeb67da2b6/LanguageExt.Core/Monads/Alternative%20Value%20Monads/Try/TryAsync/TryAsync.Extensions.cs#L486-L496). So, 
1. are the Try and TryAsync supposed to be symmetric?
2. how does one match on the type of exception inside the Error?
   (I feel like I ought to have a `ExpectedException` here but [I can't see any examples](https://github.com/search?q=repo%3Alouthy%2Flanguage-ext%20ExpectedException&type=code) of constructing one or using one.)

**Reply by NickDarvey** [reply]
Hah, I just saw

https://github.com/louthy/language-ext/blob/437171250d26a5b26c516b7c0ca94ebeb67da2b6/LanguageExt.Core/Monads/Alternative%20Value%20Monads/Try/README.md?plain=1#L4-L5

Looks like I need to relearn! I've been away far too long.

I'm guessing [this `@catch` thing](https://github.com/louthy/language-ext/blob/437171250d26a5b26c516b7c0ca94ebeb67da2b6/Samples/EffectsExamples/Menu.cs#L48) is something like what I'm looking for.  

Are there examples of catching things thrown by the uhhh implementation(?) of a runtime? For example, what's the idiomatic way to catch an (expected, not really exceptional) IOException thrown if a file is locked when invoking [WriteAllLines](https://github.com/louthy/language-ext/blob/437171250d26a5b26c516b7c0ca94ebeb67da2b6/LanguageExt.Sys/Live/FileIO.cs#L75-L79)?

**Reply by louthy** [reply]
The 'big idea' (for want of a better phrase) is to get everyone using `Error`.  [`Error` has been upgraded to support exceptional, expected, and many-errors](https://github.com/louthy/language-ext/releases/tag/v4.3.0-beta).  This allows a bit more fidelity when defining errors.

You can pre-build your errors, i..e:

```c#
public static class YourErrors
{
    public static readonly Error UserNotFound = (123, "User not found");
    public static readonly Error InvalidPath = (124, "Invalid path");
    public static readonly Error UnauthorisedAccess = (125, "File locked: unauthorised access");    
    ...
}
```
You can then use those to 'pattern-match' against with `@catch`.  So, that allows you to define the set of errors for your application, concretely, in one place and then use those `readonly` fields as a strongly-typed set of errors (easy to find usage-of, easy to rename, easy to change the error-message, easy to localise, ...).

When it comes to exceptional errors and the `Aff` and `Eff` wrappers around existing exception-throwing code, you should look to catch exceptional errors in the implementation, _as soon as possible_,  and convert to something more appropriate.  Jumping away from the built-in exceptions really is a question of pragmatism: it will take effort to map all existing exceptions to well-defined errors for your domain, you could argue that it's unnecessary overhead.  The alternative view is that by explicitly catching and mapping all exceptional errors, you take control over the expected and the exceptional.  

Anyway, you can then wrap up `File.WriteLines` like so:

```c#
public static Eff<Unit> WriteAllLines(string path, Seq<string> lines) =>
    Eff(() =>
    {
        System.IO.File.WriteAllLines(path, lines);
        return unit;
    })
    | @catch(ex => ex is UnauthorizedAccessException, YourErrors.UnauthorisedAccess)
    | @catch(ex => ex is PathTooLongException,        YourErrors.InvalidPath)
    | @catch(ex => ex is ArgumentNullException,       YourErrors.InvalidPath)
    | @catch(ex => ex is ArgumentException,           YourErrors.InvalidPath)
    | @catch(ex => ex is NotSupportedException,       YourErrors.InvalidPath);
```

**Reply by NickDarvey** [reply]
> you should look to catch exceptional errors in the implementation, as soon as possible, and convert to something more appropriate. Jumping away from the built-in exceptions really is a question of pragmatism: it will take effort to map all existing exceptions to well-defined errors for your domain, you could argue that it's unnecessary overhead. The alternative view is that by explicitly catching and mapping all exceptional errors, you take control over the expected and the exceptional.

That's exactly what I'm trying to do :) you don't need to convince me there.


>you can then wrap up File.WriteLines like so

Great example. Thanks!

As part of "catching exceptional errors in the implementation, as soon as possible", shouldn't traits like `Sys.Traits.FileIO` define these `Error`-inheriting error types and force the implementation to return them? That is, making the `System.IO` ('Live') implementation of the `FileIO` trait capture the exceptions thrown by `System.IO` and return one of `YourErrors` like in your example?



> The 'big idea' (for want of a better phrase) is to get everyone using Error. [Error has been upgraded to support exceptional, expected, and many-errors](https://github.com/louthy/language-ext/releases/tag/v4.3.0-beta).

Also, just a design question, if csharp supported discriminated unions would you let users define their own error types? Or would you continue to promote this `Error` type to avoid the proliferation of a `TError` type argument everywhere or other reasons?
(I get why it exists right now, I think, but I see a cognitive burden for myself when it comes time to handle each error where I'm needing to consider what possible `Error.Code`s I have to care about, ensuring exhaustiveness, etc.)
